File: client/src/game/engine/weaponEngagement.ts
import { randomUUID } from "@/utils/generateUUID";
import { fromLonLat, get as getProjection } from "ol/proj";
import {
  DEFAULT_OL_PROJECTION_CODE,
  NAUTICAL_MILES_TO_METERS,
} from "@/utils/constants";
import Aircraft from "@/game/units/Aircraft";
import Facility from "@/game/units/Facility";
import { Circle } from "ol/geom";
import Scenario from "@/game/Scenario";
import Weapon from "@/game/units/Weapon";
import {
  getBearingBetweenTwoPoints,
  getDistanceBetweenTwoPoints,
  getNextCoordinates,
  getTerminalCoordinatesFromDistanceAndBearing,
  randomFloat,
  randomInt,
} from "@/utils/mapFunctions";
import Airbase from "@/game/units/Airbase";
import Ship from "@/game/units/Ship";
import SimulationLogs, { SimulationLogType } from "@/game/log/SimulationLogs";

export type Target = Aircraft | Facility | Weapon | Airbase | Ship;

export function isThreatDetected(
  threat: Aircraft | Weapon,
  detector: Facility | Ship | Aircraft
): boolean {
  const projection = getProjection(DEFAULT_OL_PROJECTION_CODE);
  const detectorRangeGeometry = new Circle(
    fromLonLat([detector.longitude, detector.latitude], projection!),
    detector.getDetectionRange() * NAUTICAL_MILES_TO_METERS
  );
  return detectorRangeGeometry.intersectsCoordinate(
    fromLonLat([threat.longitude, threat.latitude], projection!)
  );
}

export function weaponCanEngageTarget(target: Target, weapon: Weapon) {
  const weaponEngagementRangeNm = weapon.getEngagementRange();
  const distanceToTargetKm = getDistanceBetweenTwoPoints(
    weapon.latitude,
    weapon.longitude,
    target.latitude,
    target.longitude
  );
  const distanceToTargetNm =
    (distanceToTargetKm * 1000) / NAUTICAL_MILES_TO_METERS;
  if (distanceToTargetNm < weaponEngagementRangeNm) {
    return true;
  }
  return false;
}

export function checkTargetTrackedByCount(
  currentScenario: Scenario,
  target: Target
) {
  let count = 0;
  currentScenario.weapons.forEach((weapon) => {
    if (weapon.targetId === target.id) count += 1;
  });
  return count;
}

export function weaponEndgame(
  currentScenario: Scenario,
  weapon: Weapon,
  target: Target,
  simulationLogs: SimulationLogs
): boolean {
  currentScenario.weapons = currentScenario.weapons.filter(
    (currentScenarioWeapon) => currentScenarioWeapon.id !== weapon.id
  );
  if (randomFloat() <= weapon.lethality) {
    if (target instanceof Aircraft) {
      currentScenario.aircraft = currentScenario.aircraft.filter(
        (currentScenarioAircraft) => currentScenarioAircraft.id !== target.id
      );
    } else if (target instanceof Facility) {
      currentScenario.facilities = currentScenario.facilities.filter(
        (currentScenarioFacility) => currentScenarioFacility.id !== target.id
      );
    } else if (target instanceof Weapon) {
      currentScenario.weapons = currentScenario.weapons.filter(
        (currentScenarioWeapon) => currentScenarioWeapon.id !== target.id
      );
    } else if (target instanceof Airbase) {
      currentScenario.airbases = currentScenario.airbases.filter(
        (currentScenarioAirbase) => currentScenarioAirbase.id !== target.id
      );
    } else if (target instanceof Ship) {
      currentScenario.ships = currentScenario.ships.filter(
        (currentScenarioShip) => currentScenarioShip.id !== target.id
      );
    }
    simulationLogs.addLog(
      weapon.sideId,
      `${weapon.name} has hit and destroyed ${target.name}`,
      currentScenario.currentTime,
      SimulationLogType.WEAPON_HIT
    );
    return true;
  }
  simulationLogs.addLog(
    weapon.sideId,
    `${weapon.name} has missed ${target.name}`,
    currentScenario.currentTime,
    SimulationLogType.WEAPON_MISSED
  );
  return false;
}

export function launchWeapon(
  currentScenario: Scenario,
  origin: Aircraft | Facility | Ship,
  target: Target,
  launchedWeapon: Weapon,
  launchedWeaponQuantity: number,
  simulationLogs: SimulationLogs
) {
  if (
    origin.weapons.length === 0 ||
    launchedWeapon.currentQuantity < launchedWeaponQuantity
  )
    return;

  for (let i = 0; i < launchedWeaponQuantity; i++) {
    const nextWeaponCoordinates = getNextCoordinates(
      origin.latitude,
      origin.longitude,
      target.latitude,
      target.longitude,
      launchedWeapon.speed
    );
    const nextWeaponLatitude = nextWeaponCoordinates[0];
    const nextWeaponLongitude = nextWeaponCoordinates[1];
    const newWeapon = new Weapon({
      id: randomUUID(),
      name: `${launchedWeapon.name} #${randomInt(1000)}`,
      sideId: origin.sideId,
      className: launchedWeapon.className,
      latitude: nextWeaponLatitude,
      longitude: nextWeaponLongitude,
      altitude: launchedWeapon.altitude,
      heading: getBearingBetweenTwoPoints(
        nextWeaponLatitude,
        nextWeaponLongitude,
        target.latitude,
        target.longitude
      ),
      speed: launchedWeapon.speed,
      currentFuel: launchedWeapon.currentFuel,
      maxFuel: launchedWeapon.maxFuel,
      fuelRate: launchedWeapon.fuelRate,
      range: launchedWeapon.range,
      route: [[target.latitude, target.longitude]],
      sideColor: launchedWeapon.sideColor,
      targetId: target.id,
      lethality: launchedWeapon.lethality,
      maxQuantity: 1,
      currentQuantity: 1,
    });
    currentScenario.weapons.push(newWeapon);
  }
  launchedWeapon.currentQuantity -= launchedWeaponQuantity;
  simulationLogs.addLog(
    origin.sideId,
    `${origin.name} launched ${launchedWeaponQuantity}x ${launchedWeapon.name} at ${target.name}`,
    currentScenario.currentTime,
    SimulationLogType.WEAPON_LAUNCHED
  );
  if (launchedWeapon.currentQuantity < 1) {
    origin.weapons = origin.weapons.filter(
      (currentOriginWeapon) => currentOriginWeapon.id !== launchedWeapon.id
    );
    simulationLogs.addLog(
      origin.sideId,
      `${origin.name} has expended all stores of ${launchedWeapon.name}`,
      currentScenario.currentTime,
      SimulationLogType.WEAPON_EXPENDED
    );
  }
}

export function weaponEngagement(
  currentScenario: Scenario,
  weapon: Weapon,
  simulationLogs: SimulationLogs
) {
  const target =
    currentScenario.getAircraft(weapon.targetId) ??
    currentScenario.getFacility(weapon.targetId) ??
    currentScenario.getWeapon(weapon.targetId) ??
    currentScenario.getShip(weapon.targetId) ??
    currentScenario.getAirbase(weapon.targetId);
  if (target) {
    const weaponRoute = weapon.route;
    if (weaponRoute.length > 0) {
      // there is a weird bug where a weapon will be teleported a vast distance if it gets too close to the target but weaponEndgame is not called, current solution is to set threshold to 1 km
      if (
        getDistanceBetweenTwoPoints(
          weapon.latitude,
          weapon.longitude,
          target.latitude,
          target.longitude
        ) < 1
      ) {
        weaponEndgame(currentScenario, weapon, target, simulationLogs);
      } else {
        const nextWeaponCoordinates = getNextCoordinates(
          weapon.latitude,
          weapon.longitude,
          target.latitude,
          target.longitude,
          weapon.speed
        );
        const nextWeaponLatitude = nextWeaponCoordinates[0];
        const nextWeaponLongitude = nextWeaponCoordinates[1];
        weapon.heading = getBearingBetweenTwoPoints(
          nextWeaponLatitude,
          nextWeaponLongitude,
          target.latitude,
          target.longitude
        );
        weapon.latitude = nextWeaponLatitude;
        weapon.longitude = nextWeaponLongitude;
      }
      weapon.currentFuel -= weapon.fuelRate / 3600;
      if (weapon.currentFuel <= 0) {
        currentScenario.weapons = currentScenario.weapons.filter(
          (currentScenarioWeapon) => currentScenarioWeapon.id !== weapon.id
        );
        simulationLogs.addLog(
          weapon.sideId,
          `${weapon.name} has run out of fuel and is no longer operational`,
          currentScenario.currentTime,
          SimulationLogType.WEAPON_CRASHED
        );
      }
    }
  } else {
    currentScenario.weapons = currentScenario.weapons.filter(
      (currentScenarioWeapon) => currentScenarioWeapon.id !== weapon.id
    );
    simulationLogs.addLog(
      weapon.sideId,
      `${weapon.name} has lost its target and is no longer operational`,
      currentScenario.currentTime,
      SimulationLogType.WEAPON_CRASHED
    );
  }
}

export function aircraftPursuit(currentScenario: Scenario, aircraft: Aircraft) {
  const target = currentScenario.getAircraft(aircraft.targetId);
  if (!target) {
    aircraft.targetId = "";
    return;
  }
  if (aircraft.weapons.length < 1) return;
  aircraft.route = [
    [
      target.latitude - 0.1 < -90 ? target.latitude : target.latitude - 0.1,
      target.longitude - 0.1 < -180 ? target.longitude : target.longitude - 0.1,
    ],
  ];
  aircraft.heading = getBearingBetweenTwoPoints(
    aircraft.latitude,
    aircraft.longitude,
    target.latitude,
    target.longitude
  );
}

export function routeAircraftToStrikePosition(
  currentScenario: Scenario,
  aircraft: Aircraft,
  targetId: string,
  strikeRadiusNm: number
) {
  const target =
    currentScenario.getFacility(targetId) ||
    currentScenario.getShip(targetId) ||
    currentScenario.getAirbase(targetId) ||
    currentScenario.getAircraft(targetId);
  if (!target) return;
  if (aircraft.weapons.length < 1) return;

  const bearingBetweenAircraftAndTarget = getBearingBetweenTwoPoints(
    aircraft.latitude,
    aircraft.longitude,
    target.latitude,
    target.longitude
  );
  const bearingBetweenTargetAndAircraft = getBearingBetweenTwoPoints(
    target.latitude,
    target.longitude,
    aircraft.latitude,
    aircraft.longitude
  );
  const strikeLocation = getTerminalCoordinatesFromDistanceAndBearing(
    target.latitude,
    target.longitude,
    (strikeRadiusNm * NAUTICAL_MILES_TO_METERS) / 1000,
    bearingBetweenTargetAndAircraft
  );

  aircraft.route.push([strikeLocation[0], strikeLocation[1]]);
  aircraft.heading = bearingBetweenAircraftAndTarget;
}

File: gym/blade/engine/weaponEngagement.py
from blade.units.Aircraft import Aircraft
from blade.units.Ship import Ship
from blade.units.Facility import Facility
from blade.units.Airbase import Airbase
from blade.units.Weapon import Weapon
from blade.Scenario import Scenario
from shapely.geometry import Point
from uuid import uuid4

from blade.utils.constants import NAUTICAL_MILES_TO_METERS
from blade.utils.utils import (
    get_bearing_between_two_points,
    get_distance_between_two_points,
    get_next_coordinates,
    get_terminal_coordinates_from_distance_and_bearing,
    random_float,
    random_int,
)

Target = Aircraft | Facility | Weapon | Airbase | Ship


def is_threat_detected(
    threat: Aircraft | Weapon, detector: Facility | Ship | Aircraft
) -> bool:
    detector_geometry = Point([detector.latitude, detector.longitude]).buffer(
        detector.get_detection_range()
        / 60  # rough conversion from nautical miles to degrees
    )
    threat_geometry = Point([threat.latitude, threat.longitude])
    return detector_geometry.contains(threat_geometry)


def weapon_can_engage_target(target: Target, weapon: Weapon) -> bool:
    weapon_engagement_range_nm = weapon.get_engagement_range()

    distance_to_target_km = get_distance_between_two_points(
        weapon.latitude, weapon.longitude, target.latitude, target.longitude
    )

    distance_to_target_nm = (distance_to_target_km * 1000) / NAUTICAL_MILES_TO_METERS

    return distance_to_target_nm < weapon_engagement_range_nm


def check_target_tracked_by_count(current_scenario: Scenario, target: Target) -> int:
    count = 0
    for weapon in current_scenario.weapons:
        if weapon.target_id == target.id:
            count += 1
    return count


def weapon_endgame(current_scenario: Scenario, weapon: Weapon, target: Target) -> bool:
    current_scenario.weapons.remove(weapon)
    if random_float(0, 1) <= weapon.lethality:
        if isinstance(target, Aircraft):
            current_scenario.aircraft.remove(target)
        elif isinstance(target, Ship):
            current_scenario.ships.remove(target)
        elif isinstance(target, Facility):
            current_scenario.facilities.remove(target)
        elif isinstance(target, Airbase):
            current_scenario.airbases.remove(target)
        elif isinstance(target, Weapon):
            current_scenario.weapons.remove(target)
        return True
    return False


def launch_weapon(
    current_scenario: Scenario,
    origin: Facility | Ship | Aircraft,
    target: Target,
    launched_weapon: Weapon,
    launched_weapon_quantity: int,
) -> None:
    if (
        len(origin.weapons) == 0
        or launched_weapon.current_quantity < launched_weapon_quantity
    ):
        return

    for _ in range(launched_weapon_quantity):
        next_weapon_coordinates = get_next_coordinates(
            origin.latitude,
            origin.longitude,
            target.latitude,
            target.longitude,
            launched_weapon.speed,
        )
        next_weapon_latitude = next_weapon_coordinates[0]
        next_weapon_longitude = next_weapon_coordinates[1]
        new_weapon = Weapon(
            id=str(uuid4()),
            name=f"{launched_weapon.name} #{random_int(0, 1000)}",
            side_id=origin.side_id,
            class_name=launched_weapon.class_name,
            latitude=next_weapon_latitude,
            longitude=next_weapon_longitude,
            altitude=launched_weapon.altitude,
            heading=get_bearing_between_two_points(
                next_weapon_latitude,
                next_weapon_longitude,
                target.latitude,
                target.longitude,
            ),
            speed=launched_weapon.speed,
            current_fuel=launched_weapon.current_fuel,
            max_fuel=launched_weapon.max_fuel,
            fuel_rate=launched_weapon.fuel_rate,
            range=launched_weapon.range,
            route=[[target.latitude, target.longitude]],
            side_color=launched_weapon.side_color,
            target_id=target.id,
            lethality=launched_weapon.lethality,
            current_quantity=1,
            max_quantity=1,
        )
        current_scenario.weapons.append(new_weapon)
    launched_weapon.current_quantity -= launched_weapon_quantity
    if launched_weapon.current_quantity < 1:
        origin.weapons.remove(launched_weapon)


def weapon_engagement(current_scenario: Scenario, weapon: Weapon) -> None:
    target = current_scenario.get_target(weapon.target_id)
    if target is None:
        current_scenario.weapons.remove(weapon)
    else:
        weapon_route = weapon.route
        if len(weapon_route) > 0:
            # there is a weird bug where a weapon will be teleported a vast distance if it gets too close to the target but weaponEndgame is not called, current solution is to set threshold to 1 km
            if (
                get_distance_between_two_points(
                    weapon.latitude,
                    weapon.longitude,
                    target.latitude,
                    target.longitude,
                )
                < 1
            ):
                weapon_endgame(current_scenario, weapon, target)
            else:
                next_weapon_coordinates = get_next_coordinates(
                    weapon.latitude,
                    weapon.longitude,
                    target.latitude,
                    target.longitude,
                    weapon.speed,
                )
                next_weapon_latitude = next_weapon_coordinates[0]
                next_weapon_longitude = next_weapon_coordinates[1]
                weapon.heading = get_bearing_between_two_points(
                    next_weapon_latitude,
                    next_weapon_longitude,
                    target.latitude,
                    target.longitude,
                )
                weapon.latitude = next_weapon_latitude
                weapon.longitude = next_weapon_longitude
                weapon.current_fuel -= weapon.fuel_rate / 3600
                if weapon.current_fuel <= 0:
                    current_scenario.weapons.remove(weapon)


def aircraft_pursuit(
    current_scenario: Scenario,
    aircraft: Aircraft,
) -> None:
    target = current_scenario.get_aircraft(aircraft.target_id)
    if target is None:
        aircraft.target_id = ""
        return
    if len(aircraft.weapons) < 1:
        return
    
    TRAIL_DISTANCE_NM = 5
    trail_km = (TRAIL_DISTANCE_NM * NAUTICAL_MILES_TO_METERS) / 1000
    behind_bearing = (target.heading + 180) % 360
    trail_position = get_terminal_coordinates_from_distance_and_bearing(
        target.latitude,
        target.longitude,
        trail_km,
        behind_bearing,
    )
    trail_latitude = trail_position[0]
    trail_longitude = trail_position[1]

    aircraft.route = [
        [trail_latitude, trail_longitude],
    ]
    aircraft.heading = get_bearing_between_two_points(
        aircraft.latitude,
        aircraft.longitude,
        trail_latitude,
        trail_longitude,
    )


def route_aircraft_to_strike_position(
    current_scenario: Scenario,
    aircraft: Aircraft,
    target_id: str,
    strike_radius_nm: float,
) -> None:
    target = current_scenario.get_target(target_id)
    if target is None:
        return
    if len(aircraft.weapons) < 1:
        return

    bearing_between_aircraft_and_target = get_bearing_between_two_points(
        aircraft.latitude, aircraft.longitude, target.latitude, target.longitude
    )
    bearing_between_target_and_aircraft = get_bearing_between_two_points(
        target.latitude, target.longitude, aircraft.latitude, aircraft.longitude
    )
    strike_location = get_terminal_coordinates_from_distance_and_bearing(
        target.latitude,
        target.longitude,
        (strike_radius_nm * NAUTICAL_MILES_TO_METERS) / 1000,
        bearing_between_target_and_aircraft,
    )

    aircraft.route.append([strike_location[0], strike_location[1]])
    aircraft.heading = bearing_between_aircraft_and_target
